### ISSUE TITLE
Grove: open a model by blob name, and give npu_open() a length and a bounded FlatBuffer verifier (#49 Step 4a)

### ISSUE THREAD

**Issue by owhinata** [owhinata/ThreadXShell#93]
## Summary

Step 2 (#92) put an asset store on the external NOR, so a file can be placed
from the console. The models are still not in it: `nn open cls|det` reads a
FIXED address that `board.cmake` compiles in, and updating one needs the host
build tree and `--target flash-model-*`. The detector cannot be committed
(licence), so on a clean tree that path starts by reproducing a pipeline.

Step 4a makes `nn open <name>` resolve a blob slot instead, and -- because a
model read out of a slot is a payload this port did not produce -- gives
`npu_open()` the two things it needs to be allowed to parse one: a LENGTH, and
a bounded FlatBuffer verifier run before anything walks the buffer.

It deliberately does NOT touch the layout. `GROVE_MODEL_*` stays, the fixed
addresses stay reachable, and the writable interval does not move. That is
Step 4b, and splitting is the point: 4b extends the writable interval up to
`0xFFE000`, which puts the OLD model region within reach of raw `nor erase` /
`nor write` -- so it must not happen while `nn` still reads a fixed address.

## Environment

- Board: Grove Vision AI V2
- Reproduced at: 9228db0

## What lands

**`npu_open()` gets a length and a verifier.** It currently range-checks only
that a few bytes remain in the window; `GetModel()` is then a cast and every
accessor follows an offset. A malformed model can reference outside the slot
and outside the window.

- `npu_open(addr, model_len, arena, arena_len)`; the length comes from the blob
  header.
- `tflite::VerifyModelBuffer()` with a bounded `flatbuffers::Verifier` BEFORE
  `GetModel()`. Order: range -> identifier -> verifier -> payload walk.
- The length needs a LOWER bound too: the identifier is read from `raw + 4`, so
  a short `model_len` would read outside the declared bounds before the
  verifier ever sees them.
- NOT the verifier's default limits (depth 64, tables 1,000,000) -- the
  generated verifier recurses and the shell stack is 4,096 B. Measured on the
  real Vela model: **depth 4, tables 19** (Vela folds the graph into one custom
  op). The shipped limits are those with headroom, stated explicitly.
- The raw form takes a length as well: `nn open --addr <addr> <len>`. Making
  "the rest of the window" the length would put the neighbouring slots and the
  bootloader's `slot-header` inside the valid reference range, which is not a
  bounds check.
- New `NPU_ERR_MODEL_FORMAT`, distinct from magic / schema / payload.

Measured cost of the verifier: **ITCM 214,368 -> 226,368 B (+12,000)**, free
47,776 -> 35,776 B, against the gate's 8,192 B headroom requirement.

**`nn open <name>` resolves a slot, without ever dropping the lease.**
`BLOB_VALID` is a statement about the HEADER only (blob_state.h), and
`blob_stat()` takes and RETURNS a lease per slot -- so a writer can move in
between resolving a name and parsing the model. The order is fixed:

1. the `nn` ownership gate (refuse when already open)
2. `npu_hw_init()` FIRST, so `NOR_LEASE_NPU` is held from here on
3. inside it, walk every slot. Candidates are VALID only, and 0 matches /
   duplicate name / BUSY / FAULT / bad MAP are classified separately rather
   than folded into "not found"
4. `blob_verify()` the chosen slot (CRC), and take the final length/address
   from the info IT returned
5. `npu_open()`
6. any failure on the way: `npu_hw_deinit()`

Needs a blob read path for a caller that already holds a lease (today there is
the lease-taking form and the reservation form), and a way to ask the NOR
whether a token is live.

**Host side keeps the full gate.** The device bounds check does not replace
`verify_vela_model` -- that one checks single subgraph, every operator on
Ethos-U, int8 I/O, offline plan, unreachable tensors, arena, BlazeFace shapes
-- and the device's check happens AFTER ~40 s of erase and transfer. So a
board-owned "verified send" wrapper becomes the official path, pointed at by
picocom's `--send-cmd`: staging copy -> `verify_vela_model` -> only on success
send that same file. `--profile cls|det` explicitly (no guessing from a
filename), verifier output to stderr (not down the YMODEM line), and
fail-closed when the host C++ compiler is absent.

## Notes

- Plan reviewed by codex over 4 rounds (3 faces LGTM).
- The negative test on hardware deliberately BYPASSES the wrapper with a raw
  `sb -k`: a CRC-valid but FlatBuffer-malformed model is the only way to reach
  the new device-side bound, and the wrapper would reject it on the host first.
- Related: #49 (epic), #92 (the store), #45 (the detector and its host gate),
  #90 (every reader of the XIP window holds a lease).

**Comment by owhinata** [comment]
Landed on `main` as `9228db0..1feed06` (`e180037` code, `1feed06` docs).

## Verified on hardware

**The bound, on a real model, with no flash write** -- the pre-#93 fixed copy is
still programmed, so the raw form checks the new length + verifier by itself:

```
nn open --addr 0x3AB7B000 1704672
nn run                              # 91 ms
```

**The blob path**, after `blob erase 0` and a `send_verified_model.sh` transfer
into slot 0:

```
stored   : 'cls' in slot 0, 1704672 B, crc32 8E679A3F
cost     : 30 NOR transaction(s), rx_drop +0
blob verify 0  -> PASS
nn open cls    -> 'cls' from slot 0 open at 0x3a201000, 1704672 B
nn run         -> 91 ms
```

| | `--addr 0x3AB7B000` | `nn open cls` (slot 0) |
|---|---|---|
| arena used | **385748 B** | **385748 B** |
| inference | 91 ms | 91 ms |
| input | int8 [1x224x224x3] scale 20324/1e6 zp -8 | same |
| output | int8 [1x10] scale 65271/1e6 zp 4 | same |

The arena figure is the evidence, not the top-5: `nn run` captures a live frame
each time, so two runs cannot produce identical logits. Arena bytes and the
tensor descriptors are deterministic, and they agree exactly -- the two copies
are the same model. Two known properties re-confirmed on the way: the host
gate's arena reads 386,096 B against the device's 385,748 (+0.09%, the 64-bit
host's 8-byte pointers, over-reporting on the safe side), and 30 NOR
transactions is the #49 Step 2 budget exactly.

**The negative test** -- a 4,096 B file with a valid `TFL3` identifier and a
root offset outside the buffer, sent with a bare `sb -k` because the wrapper
rejects it on the host:

```
stored   : 'badmodel' in slot 9, 4096 B, crc32 7DF0BB5A
blob verify 9    -> PASS, 4096 B, crc32 7DF0BB5A     <- the transfer was intact
nn open badmodel -> nn: model flatbuffer is malformed (0x3aac1000, 4096 B)
nn info          -> npu down                          <- nothing half-open
blob list        -> normal                            <- port healthy
```

This is the one path that reaches the new bound, and it is exactly the case the
CRC cannot catch: the stored checksum is of the stream that arrived.

**Refusals**, all in the command's own vocabulary rather than a status from
three layers down:

```
blob write foo 9        (nn open) -> nor: busy -- `nn` has a model open (nn close)
nor info                          -> leases : 1 (mask 0x1) -- `nn` has a model open
nn open nosuchname                -> 'nosuchname': no slot holds that name
nn open --addr 0x3AB7B000 8       -> length 8 is not between 16 and 4739072
nn open --addr 0x3AB7B000 99999999 -> length ... is not between 16 and 4739072
```

Off-script, a `blob write` with no sender started was left to time out: it
erased the slot, received nothing, and gave up cleanly (`262144 B erased, 0 B
received, 1 tx`) with the port healthy.

## Measurements

- verifier in the image: **ITCM 214,368 -> 226,368 B** (+12,000); whole change
  227,296 B, free 34,848 against the gate's 8,192 headroom
- verifier limits: the real Vela model needs **depth 4, 19 tables**; shipped 16
  and 4096
- stack: `-fstack-usage` puts every frame in the recursive cycle at **<= 24 B**,
  so a nesting level costs ~64 B. 16 levels ~1.0 KB against a 4,096 B shell
  stack; the **default depth of 64 would be ~4.1 KB**, the whole stack. That is
  why the limits are stated rather than defaulted.

## A real bug the work turned up

`verify_vela_model` was constructing its `flatbuffers::Verifier` with the
DEFAULT limits, so a model could pass the host gate and be refused by the
firmware -- on the serial line, after the erase. Both now link `npu_verify.h`
and run the same check.

## Adversarial review

No BLOCKING. Both focus questions ("can the CRC'd bytes and the parsed bytes
differ", "can a verified model reference outside its slot or the window") came
back not exploitable. Three notes, all documented rather than fixed:

- the raw `--addr` form is bounded by the window, not by a slot -- which is what
  it is for
- the verifier does not read the U55 command stream (opaque bytes; true before
  this change too -- the limit of the new check, not a regression)
- the host gate is not replaced by any of it

## For #94 (Step 4b)

Entry conditions met. The slot in use is **`cls`, slot 0, base `0x200000`,
payload `0x201000`, 1,704,672 B**. `det` is not in the store: the detector model
is not in this tree, so `nn detect` was not exercised and 4b should either
place it first or record that it was not checked.

Slot 9 still holds the 4 KB `badmodel` from the negative test.
